gtsam/sfm/ShonanAveraging.cpp

```diff
@@ -944,6 +944,36 @@ ShonanAveraging2::ShonanAveraging2(string g2oFile, const Parameters &parameters)
                                      parameters.getUseHuber()),
                          parameters) {}
 
+// Extract Rot2 measurement from Pose2 betweenfactors
+// Modeled after similar function in dataset.cpp
+static BinaryMeasurement<Rot2> convertPose2ToBinaryMeasurementRot2(
+    const BetweenFactor<Pose2>::shared_ptr &f) {
+  auto gaussian =
+      boost::dynamic_pointer_cast<noiseModel::Gaussian>(f->noiseModel());
+  if (!gaussian)
+    throw std::invalid_argument(
+        "parseMeasurements<Rot2> can only convert Pose2 measurements "
+        "with Gaussian noise models.");
+  const Matrix3 M = gaussian->covariance();
+  auto model = noiseModel::Gaussian::Covariance(M.block<1, 1>(2, 2));
```

Comment:
@dellaert or @ProfFan could either of you point me to where the ordering for the Pose2 elements is documented when you get a chance? Not sure if it's (x,y,theta) or (theta,x,y). Here I'm assuming it's (x,y,theta), but I read [here](https://github.com/borglab/gtsam/blob/43e8f1e5aeaf11890262722c1e5e04a11dbf9d75/gtsam/slam/dataset.cpp#L855) that in GTSAM the g2o order (t,R) is the opposite of GTSAM's (R,t) order, so maybe I have it wrong.


@dellaert regarding Pose3, I noticed that [here](https://github.com/borglab/gtsam/pull/817/files#diff-2681e2ffa3c440f21039b01556f84cf32000a6c0c251989f657835ee69622367R1004) the [3:6,3:6] block of the 6x6 covariance matrix is used for R:
```
  const Matrix6 M = gaussian->covariance();
  auto model = noiseModel::Gaussian::Covariance(M.block<3, 3>(3, 3));
  return BinaryMeasurement<Rot3>(f->key1(), f->key2(), f->measured().rotation(),
                                 model);
```

but [here](https://github.com/borglab/gtsam/blob/43e8f1e5aeaf11890262722c1e5e04a11dbf9d75/gtsam/slam/dataset.cpp#L885) it appears to be the opposite: the block [0:3,0:3] is pulled out for R:
```python
  const Matrix6 M = gaussian->covariance();
  return BinaryMeasurement<Rot3>(
      p.key1(), p.key2(), p.measured().rotation(),
      noiseModel::Gaussian::Covariance(M.block<3, 3>(0, 0), true));
}
```
Is that intentional or perhaps a typo? Thanks!

Reply:
Assuming that you are talking about the tangent space, it's
```c++
/* ************************************************************************* */
Vector3 Pose2::Logmap(const Pose2& p, OptionalJacobian<3, 3> H) {
  if (H) *H = Pose2::LogmapDerivative(p);
  const Rot2& R = p.r();
  const Point2& t = p.t();
  double w = R.theta();
  if (std::abs(w) < 1e-10)
    return Vector3(t.x(), t.y(), w);
  else {
    double c_1 = R.c()-1.0, s = R.s();
    double det = c_1*c_1 + s*s;
    Point2 p = R_PI_2 * (R.unrotate(t) - t);
    Point2 v = (w/det) * p;
    return Vector3(v.x(), v.y(), w);
  }
}
```

Reply:
thanks!